cmd/build/v2/smartbuild/sequential.go

```diff
@@ -0,0 +1,198 @@
+)
+
+// ImageTagger defines the interface for managing image tags and references.
+type ImageTagger interface {
+	GetGlobalTagFromDevIfNeccesary(tags, namespace, registryURL, buildHash string, ic registry.ImageCtrl) string
+	GetImageReferencesForTag(manifestName, svcToBuildName, tag string) []string
+	GetImageReferencesForDeploy(manifestName, svcToBuildName string) []string
+}
+
+// SmartBuildController defines the interface for smart build operations.
+type SmartBuildController interface {
+	GetBuildHash(buildInfo *build.Info, service string) string
+	CloneGlobalImageToDev(globalImage, svcImage string) (string, error)
+}
+
+type CacheProbe interface {
+	IsCached(manifestName, image, buildHash, svcToBuild string) (bool, string, error)
+	LookupReferenceWithDigest(reference string) (string, error)
+	GetFromCache(svc string) (hit bool, reference string)
+}
+
+type ServiceEnvVarsSetter interface {
+	SetServiceEnvVars(service, reference string)
+}
+
+type SequentialCheckStrategy struct {
+	tagger               ImageTagger
+	hasher               hasherController
+	imageCacheChecker    CacheProbe
+	ioCtrl               *io.Controller
+	serviceEnvVarsSetter ServiceEnvVarsSetter
+	cloner               *cloner
+}
+
+func NewSequentialCheckStrategy(tagger ImageTagger, hasher hasherController, imageCacheChecker CacheProbe, ioCtrl *io.Controller, serviceEnvVarsSetter ServiceEnvVarsSetter, cloner *cloner) *SequentialCheckStrategy {
+	return &SequentialCheckStrategy{
+		tagger:               tagger,
+		hasher:               hasher,
+		imageCacheChecker:    imageCacheChecker,
+		ioCtrl:               ioCtrl,
+		serviceEnvVarsSetter: serviceEnvVarsSetter,
+		cloner:               cloner,
+	}
+}
+
+func (s *SequentialCheckStrategy) CheckServicesCache(ctx context.Context, manifestName string, buildManifest build.ManifestBuild, svcsToBuild []*buildTypes.BuildInfo) (cachedSvcs []*buildTypes.BuildInfo, notCachedSvcs []*buildTypes.BuildInfo, err error) {
+	// svcsToBuild is already ordered by dependencies (from DAG.Ordered())
+	// so we can optimize by stopping the check as soon as we find a service that's not cached.
+	// All subsequent services that depend on it will also need to be rebuilt.
+
+	dependantMap := make(map[string][]string)
+	// Build dependantMap: for each service, track which services depend on it
+	for svcName, buildInfo := range buildManifest {
+		for _, dep := range buildInfo.DependsOn {
+			dependantMap[dep] = append(dependantMap[dep], svcName)
+		}
+	}
+
+	// Track processed services to avoid duplicates
+	processed := make(map[string]bool)
+
+	// Map of service name to BuildInfo for quick lookup
+	buildInfoByName := make(map[string]*buildTypes.BuildInfo, len(svcsToBuild))
+	for _, bi := range svcsToBuild {
+		buildInfoByName[bi.Name()] = bi
+	}
+
+	for _, svc := range svcsToBuild {
+		// Skip if already processed
+		name := svc.Name()
+		if processed[name] {
+			continue
+		}
+
+		s.ioCtrl.SetStage(fmt.Sprintf("Building service %s", svc))
```

Comment:
As discussed offline, I would create a new stage to represent the smart build check. As it is now, it would be very error prone because if we print something when the image is not cached, it could happen that we send messages to stages created previously and the experience is not good in the UI.

I would create a stage like `Checking if images were already built`, where we print the message for the images cached (we could even consider to print a messages with images that will be built).

And then, we have one stage `Building service ...` per image that needs to be built